core/arch/arm/mm/tee_pager.c

```diff
@@ -1303,37 +1379,151 @@ static void make_dirty_page(struct tee_pager_pmem *pmem,
+		 * to another page.
+		 */
+		pmem = pager_spare_pmem;
+		pager_spare_pmem = NULL;
+		pmem_assign_fobj_page(pmem, area, page_va);
+
+		if (writable)
+			pmem->flags |= PMEM_FLAG_DIRTY;
+
+		pager_deploy_page(pmem, area, page_va,
+				  false /*!clean_user_cache*/, writable);
+	} else if (writable && !(attr & TEE_MATTR_PW)) {
+		pmem = pmem_find(area, tblidx);
+		/* Note that pa is valid since TEE_MATTR_VALID_BLOCK is set */
+		make_dirty_page(pmem, area, tblidx, pa, page_va);
 	}
+}
 
-	TAILQ_REMOVE(&tee_pager_pmem_head, pmem, link);
-	pmem_clear(pmem);
-	if (at == PAGER_AREA_TYPE_LOCK) {
-		/* Move page to lock list */
-		if (tee_pager_npages <= 0)
-			panic("running out of page");
-		tee_pager_npages--;
-		set_npages();
-		TAILQ_INSERT_TAIL(&tee_pager_lock_pmem_head, pmem, link);
-	} else {
-		/* move page to back */
-		TAILQ_INSERT_TAIL(&tee_pager_pmem_head, pmem, link);
+static void pager_get_page(struct tee_pager_area *area, struct abort_info *ai,
+			   bool clean_user_cache)
+{
+	vaddr_t page_va = ai->va & ~SMALL_PAGE_MASK;
+	unsigned int tblidx = area_va2idx(area, page_va);
+	struct tee_pager_pmem *pmem = NULL;
+	bool writable = false;
+	uint32_t attr = 0;
+
+	/*
+	 * Get a pmem to load code and data into, also make sure
+	 * the corresponding IV page is available.
+	 */
+	while (true) {
+		pmem = TAILQ_FIRST(&tee_pager_pmem_head);
+		if (!pmem) {
+			EMSG("No pmem entries");
+			abort_print(ai);
+			panic();
+		}
+
+		if (pmem->fobj) {
+			pmem_unmap(pmem, NULL);
+			if (pmem_is_dirty(pmem)) {
+				uint8_t *va = pmem->va_alias;
+
+				make_iv_available(pmem->fobj, pmem->fobj_pgidx,
+						  true /*writable*/);
+				asan_tag_access(va, va + SMALL_PAGE_SIZE);
+				if (fobj_save_page(pmem->fobj, pmem->fobj_pgidx,
+						   pmem->va_alias))
+					panic("fobj_save_page");
+				asan_tag_no_access(va, va + SMALL_PAGE_SIZE);
+
+				pmem_clear(pmem);
+
+				/*
+				 * If the spare pmem was used. We need to
+				 * replace it with the just freed pmem.
+				 */
+				if (!pager_spare_pmem) {
+					TAILQ_REMOVE(&tee_pager_pmem_head,
+						     pmem, link);
+					pager_spare_pmem = pmem;
+					pmem = NULL;
+				}
+
+				/*
+				 * Check if the needed virtual page was
+				 * made available as a side effect of the
+				 * call to make_iv_available() above. If so
+				 * we're done.
+				 */
+				area_get_entry(area, tblidx, NULL, &attr);
+				if (attr & TEE_MATTR_VALID_BLOCK)
+					return;
+
+				/*
+				 * The freed pmem was used to replace the
+				 * consumed pager_spare_pmem above. Restart
+				 * to get find another pmem.
+				 */
+				if (!pmem)
+					continue;
+			}
+		}
+
+		TAILQ_REMOVE(&tee_pager_pmem_head, pmem, link);
+		pmem_clear(pmem);
+
+		pmem_assign_fobj_page(pmem, area, page_va);
+		make_iv_available(pmem->fobj, pmem->fobj_pgidx,
+				  false /*!writable*/);
+		if (pager_spare_pmem)
+			break;
+
+		/*
+		 * The spare pmem was used. We need to replace it with the
+		 * just freed pmem. And get another pmem.
```

Comment:
maybe nitpicking but I hard understood this. I think the confusion came from `just freed`, while the _just freed_ pmem has been assigned to our request virtual page, so it's not really free, even if not really used (since we're still in the page fault handler).

Actually this spare pmem management confused me. Not easy to follow. But I guess the whole pager is not easy to follow :yum:.

Maybe slightly rephrase (or maybe not, at your will):
```
/*
 * The spare pmem was used. We need to replace it. Let's use the one we
 * found and get another pmem for the requested virtual page.
 */
```

Reply:
Yeah, I'm sorry it's more complicated than ever. It's the problem with the pager accessing paged pages not directly related to the current fault address.